crates/pallet-domains/src/block_tree.rs

```diff
@@ -0,0 +1,700 @@
+//! Domain block tree
+
+use crate::{
+    BlockTree, Config, DomainBlocks, ExecutionInbox, ExecutionReceiptOf, HeadReceiptNumber,
+};
+use codec::{Decode, Encode};
+use frame_support::{ensure, PalletError};
+use scale_info::TypeInfo;
+use sp_core::Get;
+use sp_domains::v2::ExecutionReceipt;
+use sp_domains::{DomainId, OperatorId};
+use sp_runtime::traits::{CheckedSub, One, Saturating, Zero};
+use sp_std::cmp::Ordering;
+use sp_std::vec::Vec;
+
+/// Block tree specific errors
+#[derive(TypeInfo, Encode, Decode, PalletError, Debug, PartialEq)]
+pub enum Error {
+    InvalidExtrinsicsRoots,
+    UnknownParentBlockReceipt,
+    BuiltOnUnknownConsensusBlock,
+    InFutureReceipt,
+    PrunedReceipt,
+    BadGenesisReceipt,
+    UnexpectedReceiptType,
+    MaxHeadDomainNumber,
+}
+
+#[derive(TypeInfo, Debug, Encode, Decode, Clone, PartialEq, Eq)]
+pub struct DomainBlock<Number, Hash, DomainNumber, DomainHash, Balance> {
+    /// The full ER for this block.
+    pub execution_receipt: ExecutionReceipt<Number, Hash, DomainNumber, DomainHash, Balance>,
+    /// A set of all operators who have committed to this ER within a bundle. Used to determine who to
+    /// slash if a fraudulent branch of the `block_tree` is pruned.
+    ///
+    /// NOTE: there may be duplicated operator id as an operator can submit multiple bundles with the
+    /// same head receipt to a consensus block.
+    pub operator_ids: Vec<OperatorId>,
+}
+
+/// The type of receipt regarding to its freshness
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) enum ReceiptType {
+    // New head receipt that extend the longest branch
+    NewHead,
+    // Receipt that comfirms the current head receipt
+    CurrentHead,
+    // Receipt that creates a new branch of the block tree
+    NewBranch,
+    // Receipt that is newer than the head receipt but does not extend the head receipt
+    InFuture,
+    // Receipt that comfirm a non-head receipt
+    Stale,
+    // Receipt that already been pruned
+    Pruned,
+}
+
+/// Get the receipt type of the given receipt based on the current block tree state
+pub(crate) fn execution_receipt_type<T: Config>(
+    domain_id: DomainId,
+    execution_receipt: &ExecutionReceiptOf<T>,
+) -> ReceiptType {
+    let receipt_number = execution_receipt.domain_block_number;
+    let head_receipt_number = HeadReceiptNumber::<T>::get(domain_id);
```

Comment:
I dont understant this. IIUC `head_receipt_number` should be the head of the fork this receipt is part of. But head receipt number you are using may or may not be the same

Reply:
The `HeadReceiptNumber` is applied to the whole block tree not any specific fork in the tree, refer https://github.com/subspace/subspace/pull/1650#discussion_r1262581662.